src/Control/Controller.php

```diff
@@ -690,14 +697,31 @@ public static function join_links($arg = null)
             }
             if ((is_string($arg) && $arg) || is_numeric($arg)) {
                 $arg = (string) $arg;
-                if ($result && substr($result ?? '', -1) != '/' && $arg[0] != '/') {
-                    $result .= "/$arg";
+                if (strlen($result) == 0) {
+                    $result .= $arg;
+                } elseif (substr($result, -1) == '/') {
+                    $result .= trim($arg, '/');
                 } else {
-                    $result .= (substr($result ?? '', -1) == '/' && $arg[0] == '/') ? ltrim($arg, '/') : $arg;
+                    $result .= '/' . trim($arg, '/');
                 }
             }
         }
 
+        // add trailing slash if enabled and url does not contain a file extension
+        if (static::config()->get('add_trailing_slash') === true
+            && substr($result, -1) !== '/'
+            && !preg_match('/^(.*)\.([A-Za-z][A-Za-z0-9]*)$/', $result)
```

Comment:
Is there actually some rule saying file extensions start with a letter and can only contain letters or numbers?

Reply:
No idea. I got the regex from https://github.com/silverstripe/silverstripe-framework/blob/4/src/Control/HTTPRequest.php#L182.

Reply:
Given that there can be anything after the period, lets not be restrictive here.